### PR TITLE
[WIP] Initial support for CSI common

### PR DESCRIPTION
**Pull request by deitch**
**What this PR does / why we need it**:

Extracts the common CSI parts from #626 so that only the alicloud-specific parts need to be there, and we can reuse the common CSI structure for other providers.

This is successful if it becomes easier to add other providers with CSI requirements.

**Special notes for your reviewer**:
This is very much a work in progress (WIP). First cut just pulled the CSI-specific code out of #626 , and I am sure I missed some things.

The most important open question is how this would be consumed. I already have a work in progress document describing how to add a new cloud provider, it is missing this part. So, given this:

1. How would we add alicloud's CSI, or Packet's, or DigitalOcean's, or even AWS/Azure/GCP once they are out of tree? What would it look like?
2. The specific changes from #626 that were _not_ included are listed below. Do we include them but parametrize? Are they a separate chart?

Here is what is provider-specific, and thus was not pulled in from the other PR:

* In `charts/shoot-core/values.yaml`, 3 of the common images are included, but the provider-specific one is not, see [here](https://github.com/gardener/gardener/pull/626/files#diff-d34d9a2b7a397eb72b1701d2e78b2725R42)
* the entire DaemonSet for the cloud-provider-specific CSI plugin, see [here](https://github.com/gardener/gardener/pull/626/files#diff-ddb52d7f57a2d98f04e33d92e954c9ff). I think it would be good if we can have a single DaemonSet in the CSI chart for this, and just have the right parameters passed to it, i.e. the name of the image (which is different for Alicloud than for Packet than for DigitalOcean than for AWS etc.), and potentially command-line options. Not sure how to do it, though, or if we are better off just leaving it as a separate chart, e.g. `charts/csi-alicloud`, `charts/csi-packet`, etc.

cc @rfranzke 

**Release note**:

```improvement operator
Add common CSI support
```

### CODE REVIEW

**Comment by rfranzke** [comment]
Alicloud PR #626 is merged. Please rebase. @jia-jerry please take a look as well on this one.

> How would we add alicloud's CSI, or Packet's, or DigitalOcean's, or even AWS/Azure/GCP once they are out of tree? What would it look like?

Not sure yet in detail, but it will not be part of this repository (Gardener core). Instead, your out-of-tree packet botanist will deploy the CSI. Probably no common charts will be usable then.

> The specific changes from #626 that were not included are listed below. Do we include them but parametrize? Are they a separate chart?

@jia-jerry can you please assist here?

**Comment by deitch** [comment]
> Alicloud PR #626 is merged. Please rebase

I rebased and redid this, extracting the common info from `charts/csi-alicloud/` to `charts/csi/`, and left (mostly) the unique alicloud stuff in `charts/csi-alicloud/`. That having been said, it should be possible to parametrize it. But someone who understands it better can look?

**Comment by jia-jerry** [comment]
@deitch , thank you for your proposal for refactoring CSI components. I don't think it is needed by now to abstract the CSI charts into a common one by now. 
1. As @rfranzke said, we are refactoring the cloud extentions (out of tree solution). CSI charts will be maintained separately in different extention. From this kind of view, it seems that we do not need a common CSI charts.
2. Continue to 1. even we abstracted the provisioner, attacher and authorization settings, it is only used in the current extention for the current cloud provider. Why do we need to maintain CSI related components in 2 charts?
3. Also coninue to 1, even we will not use folder like **csi-alicloud**, **csi** will be OK because the currently app is only for a single cloud provider. 

The common stuff will only be considered in Gardener project, I think. @rfranzke please correct me if I made misleading comments.

**Comment by deitch** [comment]
@jia-jerry , are you saying that _nothing_ in here (or nearly nothing) is common for others? E.g. when AWS goes out of tree for EBS, and only uses csi, will there be no overlap here?

**Comment by jia-jerry** [comment]
> @jia-jerry , are you saying that _nothing_ in here (or nearly nothing) is common for others? E.g. when AWS goes out of tree for EBS, and only uses csi, will there be no overlap here?

They have some thing like provisioner and attacher in common, for csi v0.3.0, as far as I know. But the background is that we are doing refactoring. Please refer to #308. Each kind of infrastructure has its own controller implementation. This separate implementation of cause will include a dedicated csi chart. So why do we need a csi common chart? Please be noted each controller will be in a separate repo. @rfranzke Please correct me if I made misunderstanding on the extension solution.

You may consider a micro service architecture. The only thing you care about is to follow the interface contract. You don't need to abstract your module into a common module always. Where do you put this module, in another repo? If no, how do we share the common mudule?

**Comment by deitch** [comment]
No objection to it being separate. Actually, if it really is separate - where Alicloud is in its own repo, and Packet in its own, and AWS in its own, etc. - then it makes perfect sense. This is how csi itself is implemented.

**Comment by jia-jerry** [comment]
> No objection to it being separate. Actually, if it really is separate - where Alicloud is in its own repo, and Packet in its own, and AWS in its own, etc. - then it makes perfect sense. This is how csi itself is implemented.

csi are set of components that can work together. As about a chart, they usually are a set of depensive configurations. 

If you want install a service, you only need to install from a chart, which includes recursive dependencies. 

But for csi, it is not separable. The common practice is to either put all configurations into one chart or maintain a set **depensive** charts. I don't think it is a good choice to make them into isolated charts.

File: charts/shoot-core/charts/csi/charts/utils-templates/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
```
Comment (rfranzke) [inline charts/shoot-core/charts/csi/charts/utils-templates/Chart.yaml]: Should not be copied but symlinked.

**Comment by rfranzke** [comment]
Once we have progressed with extensibility and moved out the cloud specific parts there is no common chart anymore - every cloud provider will have its own CSI deployments.
With this PoV, I get @jia-jerry's point. It's probably not worth the effort splitting out comming parts now because it will
a) be not like this in the future
b) be harder/more effort to move out the cloud provider specifics with #308.

Summing up, @deitch would you agree to not invest in that now? Instead, simply add your CSI charts that are required for Packet (although there might be  some overlapping parts with Alicloud right now), but it will be easier to extract these specifics later.

**Comment by deitch** [comment]
> would you agree to not invest in that now?

Sure, makes sense. I am rather looking forward to getting the Packet driver done.
